### PR TITLE
8323688: C2: Fix UB of jlong overflow in PhaseIdealLoop::is_counted_loop()

### PR DESCRIPTION
**Pull request by chhagedorn**
The computation of `final_correction` in `PhaseIdealLoop::is_counted_loop()` could overflow which is UB:

https://github.com/openjdk/jdk/blob/dc4fd896289db1d2f6f7bbf5795fec533448a48c/src/hotspot/share/opto/loopnode.cpp#L1958-L1967

`canonicalized_correction` equals `max_int - 1` if stride is `max_int`. `limit_correction` is at most `max_int - 1` in that case. Adding both together will overflow. I don't think that any compiler would wrongly optimize this and we have not observed any issues with that. But we should still fix this UB.

The fix I propose is to simply bail out with very large positive and negative strides such that we avoid an over- or underflow with the existing logic (see added comments for how the upper bound for the stride is determined). These large strides should be very uncommon in practice and even if we encounter these, the loop would only run for a few iterations. So, a bailout seems fine. This bailout has the additional benefit that we avoid other possibly unknown issues or issues in the future with counted loops having large edge-case strides like `min_int`.

Thanks,
Christian

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8323688](https://bugs.openjdk.org/browse/JDK-8323688): C2: Fix UB of jlong overflow in PhaseIdealLoop::is_counted_loop() (**Bug** - P4)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20828/head:pull/20828` \
`$ git checkout pull/20828`

Update a local copy of the PR: \
`$ git checkout pull/20828` \
`$ git pull https://git.openjdk.org/jdk.git pull/20828/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20828`

View PR using the GUI difftool: \
`$ git pr show -t 20828`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20828.diff">https://git.openjdk.org/jdk/pull/20828.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20828#issuecomment-2325812738)

### CODE REVIEW

**Review by TobiHartmann — APPROVED** [review]
Looks good to me otherwise.

File: src/hotspot/share/opto/loopnode.cpp [hunk 1]
```diff
@@ -1954,6 +1970,10 @@ bool PhaseIdealLoop::is_counted_loop(Node* x, IdealLoopTree*&loop, BasicType iv_
   //            there is no overflow of the iv phi after the first iteration. In this case, we don't need to check (ii)
   //            again and can skip the predicate.
 
+  // Check (vi) and bail out if the stride is too big.
+  if (stride_con == min_signed_integer(iv_bt) || ABS(stride_con) > max_signed_integer(iv_bt) / 2) {
```
[the remark this answers is not in the document]
Comment (chhagedorn) [reply src/hotspot/share/opto/loopnode.cpp]: Thanks for the review, good point!

**Review by TobiHartmann — APPROVED** [review]

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by chhagedorn** [comment]
Thanks Tobias and Vladimir for your reviews!

/integrate

**Pull request closed without merging**